### PR TITLE
Add colorful profiling

### PR DESCRIPTION
**Pull request by benson31**
Just ports and C++-izes some of the profiling stuff from LBANN

### CODE REVIEW

File: include/El/core/Profiling.hpp [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+
+    std::string desc_;
+};// struct ProfileRegion
+
+inline ProfileRegion MakeProfileRegion(std::string desc, Color color)
+{
+    return ProfileRegion(std::move(desc), color);
+}
+
+// Synchronous RAII profiler.
```
Comment (ndryden) [inline include/El/core/Profiling.hpp]: It might be good to document what makes this "synchronous", since it's not obvious.

**Review by ndryden — APPROVED** [review]
Looks good to me. It might also be helpful to add the start/stop methods as an option ([here](https://docs.nvidia.com/cuda/profiler-users-guide/index.html#nvtx-range-start-stop)), for stuff that isn't nested.

**Review by timmoon10 — APPROVED** [review]
LGTM

File: include/El/core/Profiling.hpp (hunk 1 above)
Comment (benson31) [reply include/El/core/Profiling.hpp]: Done.

**Pull request merged**